### PR TITLE
 Add fix and tests for float32 pi/2 Latitude

### LINKED ISSUE #13708: float32 representation of pi/2 is rejected by `Latitude`
### Description

The closest float32 value to pi/2 is by accident slightly larger than pi/2:

```
In [5]: np.pi/2
Out[5]: 1.5707963267948966

In [6]: np.float32(np.pi/2)
Out[6]: 1.5707964
```

Astropy checks using float64 precision, rejecting "valid" alt values (e.g. float32 values read from files):

```

In [1]: from astropy.coordinates import Latitude

In [2]: import numpy as np

In [3]: lat = np.float32(np.pi/2)

In [4]: Latitude(lat, 'rad')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [4], line 1
----> 1 Latitude(lat, 'rad')

File ~/.local/lib/python3.10/site-packages/astropy/coordinates/angles.py:564, in Latitude.__new__(cls, angle, unit, **kwargs)
    562     raise TypeError("A Latitude angle cannot be created from a Longitude angle")
    563 self = super().__new__(cls, angle, unit=unit, **kwargs)
--> 564 self._validate_angles()
    565 return self

File ~/.local/lib/python3.10/site-packages/astropy/coordinates/angles.py:585, in Latitude._validate_angles(self, angles)
    582     invalid_angles = (np.any(angles.value < lower) or
    583                       np.any(angles.value > upper))
    584 if invalid_angles:
--> 585     raise ValueError('Latitude angle(s) must be within -90 deg <= angle <= 90 deg, '
    586                      'got {}'.format(angles.to(u.degree)))

ValueError: Latitude angle(s) must be within -90 deg <= angle <= 90 deg, got 90.00000250447816 deg
```

### Expected behavior

Be lenient? E.g. only make the comparison up to float 32 precision?

### Actual behavior
See error above

### Steps to Reproduce

See snippet above.

### System Details

```
Linux-5.15.65-1-MANJARO-x86_64-with-glibc2.36
Python 3.10.7 (main, Sep  6 2022, 21:22:27) [GCC 12.2.0]
Numpy 1.23.3
pyerfa 2.0.0.1
astropy 5.0.1
Scipy 1.9.1
Matplotlib 3.5.2
```

### PR DESCRIPTION
**Pull request by maxnoe**
### Description

This adds test cases and for now the easy part (everything float32) for the `Latitude._validate_angle` method as discussed here:

(Hopefully soon) Fixes #13708 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/coordinates/tests/test_angles.py
Comment (maxnoe) [inline astropy/coordinates/tests/test_angles.py:1102]: This and the case below are still failing, since I didn't yet implement the coercion of these values.

File: astropy/coordinates/tests/test_angles.py
Comment (dhomeier) [inline astropy/coordinates/tests/test_angles.py:1103]:
```suggestion
        (np.float32(np.pi / 2), np.float(np.pi / 2), None),
```
This will (and does successfully) create an `Angle` of the input dtype, so you need to compare against that.

**Review by dhomeier — CHANGES_REQUESTED** [review]
I had half-expected the second test to pass already, and in fact it does half-succeed. ;-)

I've tested that even checking for `np.any(angles.value < -np.float32(limit))` etc. should work for the relaxed test, at least for float32 `value` directly cast to float64; that might not work e.g. for values read in from ascii with insufficient number of decimals, but I don't think we need to care about that kind of imprecise input.

File: astropy/coordinates/angles.py
Comment (mhvk) [inline astropy/coordinates/angles.py:590]: for speed purposes, `limit = self.dtype.type(0.5 * np.pi)` seems fastest (even for unchanged dtype)

**Review by mhvk — COMMENTED** [review]
My sense would be to get in the simple version, which is definitely a bug fix.  I am not quite sold on the coercion.

**Comment by dhomeier** [comment]
My main concern is that if the simple fix is merged, the fix for the full scope gets forgotten. Although one could also take the position that users are responsible for their own input here (i.e. when casting to float64 at any prior stage ensure they stay within bounds). Perhaps a remark in the `Latitude` section of `angles.rst` would be enough.

**Comment by mhvk** [comment]
My own sense is indeed that it is OK to leave this to the user -- we have had no complaints until now, so even the actual bug is not too common (it may well be that most data is in degrees, in which case the limit is identical in float32 and float64). It still might be good to add some documentation; not sure whether the general docs or the docstring would be best; perhaps arguing for the docstring is that in the general docs one has to be sure not to distract with information that is not useful for most people.

File: astropy/coordinates/angles.py
Comment (maxnoe) [reply astropy/coordinates/angles.py]: done

**Review by mhvk — DISMISSED** [review]
Great!

**Comment by mhvk** [comment]
Oops, I approved, but the one things still missing is a changelog entry - since the case now fixed definitely was a bug!

**Comment by maxnoe** [comment]
Changelog entry added

File: astropy/coordinates/tests/test_angles.py
Comment (eerovaher) [inline astropy/coordinates/tests/test_angles.py:1138]: It would be good to test the error message here, and the expected error message would also describe what this test is meant for so that the docstring would not be required.

File: astropy/coordinates/tests/test_angles.py
Comment (eerovaher) [inline astropy/coordinates/tests/test_angles.py:1113]: I'd prefer if the docstring here would be more explicit so that its purpose could be understood even without looking up the issue.

File: astropy/coordinates/tests/test_angles.py
Comment (maxnoe) [reply astropy/coordinates/tests/test_angles.py]: done

File: astropy/coordinates/tests/test_angles.py
Comment (maxnoe) [reply astropy/coordinates/tests/test_angles.py]: done

**Review by mhvk — APPROVED** [review]
OK, thanks!

**Comment by mhvk** [comment]
I think this is all OK, so approved. But I cannot merge without dismissing @dhomeier's request for changes, so rather let him have a last look.

File: astropy/coordinates/tests/test_angles.py
Comment (dhomeier) [inline astropy/coordinates/tests/test_angles.py:1136]: This should in fact be far more sensitive (to at least 0.50000000001), but for this particular bug may be sufficient.

**Comment by dhomeier** [comment]
Thanks; my change request is in fact fully addressed by the changes to the test!

Still thinking about documenting the second case, though as @mhvk noted this appears to have rarely been a problem so far, but
> arguing for the docstring is that in the general docs one has to be sure not to distract with information that is not useful for most people.

Following the
```python
-90.0 * u.deg <= angle(s) <= +90.0 * u.deg

    Any attempt to set a value outside that range will result in a `ValueError`.
```
one might add something like
```
Note that this condition will be evaluated within the precision of the input angle `dtype`.
```
(Should we add `equivalent to -0.5 * np.pi * u.deg <= angle <= +0.5 * np.pi * u.rad`?)

**Comment by mhvk** [comment]
Yes, that location makes sense to me, and I can see that as part of the bug fix. But your suggested text states only something that ideally would have been the case already. An alternative would be to move the `ValueError` to the `Raises` section and be more explicit. I.e., 
```
Raises
------
ValueError
    If any of the input values is outside the valid range.
    Note that this is evaluated at the input value dtype,
    and may be susceptible to rounding errors if values
    are read in lower precision and then upcast. For
    instance, ``Latitude(float(np.float32(np.pi/2)), 'rad')``
    will raise an exception, since the 32-bit float value
    of pi/2 is slightly larger than the 64-bit value.
```
Still wonder if this is not too much, so can go with something like this or just merging as is.

**Review by dhomeier — APPROVED** [review]

**Comment by mhvk** [comment]
Actually, another thought: the extra text could also be part of the exception text -- that at least has the best hope of being seen by an unwary user!  But again, I'm also happy to merge as it is now.

File: astropy/coordinates/tests/test_angles.py
Comment (pllim) [inline astropy/coordinates/tests/test_angles.py:1098]: Do you think this would benefit from using hypothesis or is that overkill?

Comment (mhvk) [reply astropy/coordinates/tests/test_angles.py:1098]: Overkill... It might well find little corner cases that I'd rather be kept hidden...

Comment (maxnoe) [reply astropy/coordinates/tests/test_angles.py:1098]: 🙈

**Comment by mhvk** [comment]
Let's get this in! thanks @maxnoe!

**Comment by pllim** [comment]
Might be the isort wrecking havoc. Can you please do a manual backport? Thanks! 🙏

File: astropy/coordinates/angles.py
Comment (pllim) [inline astropy/coordinates/angles.py:590]: I am not sure if we can backport this to v5.0.x . This is the conflict I am seeing w.r.t. v5.0.x branch.

```
    def _validate_angles(self, angles=None):
        """Check that angles are between -90 and 90 degrees.
        If not given, the check is done on the object itself"""
        # Convert the lower and upper bounds to the "native" unit of
        # this angle.  This limits multiplication to two values,
        # rather than the N values in `self.value`.  Also, the
        # comparison is performed on raw arrays, rather than Quantity
        # objects, for speed.
        if angles is None:
            angles = self
<<<<<<< HEAD
        lower = u.degree.to(angles.unit, -90.0)
        upper = u.degree.to(angles.unit, 90.0)
=======

        if angles.unit is u.deg:
            limit = 90
        elif angles.unit is u.rad:
            limit = self.dtype.type(0.5 * np.pi)
        else:
            limit = u.degree.to(angles.unit, 90.0)

>>>>>>> a3a25b93d... Merge pull request #13745 from maxnoe/pihalf
```

Comment (pllim) [reply astropy/coordinates/angles.py:590]: Same problem with v5.1.x

Comment (pllim) [reply astropy/coordinates/angles.py:590]: Looks like https://github.com/astropy/astropy/pull/13498 needs to be backported too if we want this backported. If I don't hear back in a few days, I am going to change the milestone to v5.2 for this PR and not do any backports.

**Comment by pllim** [comment]
Update: I changed the milestone. This patch will *not* be backported.

File: astropy/coordinates/angles.py
Comment (dhomeier) [reply astropy/coordinates/angles.py:590]: Does it? Why not simply put `upper = self.dtype.type(u.degree.to(angles.unit, 90.0))` etc.?
Should not make a difference for `u.deg` as 90 is exact in both float32 and float 64, right?
And for any other units it should do no harm (might in fact have made sense to apply it to the `else` case here as well).

Comment (maxnoe) [reply astropy/coordinates/angles.py:590]: Yes, that should be it, I can open the backport pr.

**Comment by pllim** [comment]
Update: Manual backport attempt with a slightly different fix at #13803 (targeted at v5.0.x but an attempt to forward port to v5.1.x will be made on merge). When that is settled, I guess I can change the milestone back.

**Pull request merged**
